ads/google/utils.js

```diff
@@ -14,16 +14,289 @@
+
+/**
+ * @param {number} num Number, non-negative.
+ * @param {number} digits Number of digits, max 20.
+ * @return {string}
+ */
+function formatFixedWidthInteger(num, digits) {
+  const intPart = String(Math.max(Math.round(num), 0));
+  const len = intPart.length;
+  digits = Math.min(digits, 20);
+  if (len > digits) {
+    return '99999999999999999999'.substr(0, digits);
+  }
+  return '00000000000000000000'.substr(0, digits - len) + intPart;
+}
+
+function adKey(slotNumber, slot, viewport, screen) {
+  // Assume Nexus 6 screen, if we don't get either a viewport or a screen.
+  const viewWidth = viewport ? viewport.width : screen ? screen.width : 1080;
+  const viewHeight = viewport ? viewport.height : screen ? screen.height : 1920;
+  const slotTop = slot.y + (viewport ? viewport.y : 0);
+  const slotLeft = slot.x + (viewport ? viewport.x : 0);
+  return formatFixedWidthInteger(slotNumber, 2) +
+    // ad slot top, in 1/5 viewport height units
+    formatFixedWidthInteger(slotTop * 5 / viewHeight, 4) +
+    // ad slot left, in 1/5 viewport width units
+    formatFixedWidthInteger(slotLeft * 5 / viewWidth, 4);
+}
+
+/**
+ * Make the URL for an ad request, from an amp-ad element.
+ * @param {number} slotNumber
+ * @param {!Window} global
+ * @param {!Object} data
+ * @param {!IntersectionObserverEntry} intersectionRecord
+ * @return {string}
+ */
+export function adsenseRequestUrlForAmpAd(slotNumber, global, data,
+                                          intersectionRecord) {
+  // Assumes that data has been checked for valid parameters, via checkData.
+  return adsenseRequestUrl(global.context.startTime, slotNumber, global,
+                           data, 1, global.context.canonicalUrl,
+                           getCorrelator(global), intersectionRecord);
+}
+
+/**
+ * @private {number}
+ */
+const MAX_URL_LENGTH_ = 4096;
+
+/**
+ * @private {number}
+ */
+const MIN_USEFUL_URL_PARAM_LENGTH_ = '&url=http%3A%2F%2F'.length + 1;
+
+/**
+ * @private {string}
+ */
+const TRUNC_PARAM_ = '&trunc=1';
+
+/**
+ * Make the URL for an ad request.
+ * @param {number} startTime
+ * @param {number} slotNumber
+ * @param {!Window} global
+ * @param {!Object} data
+ * @param {number) isAmp
+ * @param {string} canonicalUrl
+ * @param {number} correlator
+ * @param {!IntersectionObserverEntry} intersectionRecord
+ * @return {string}
+ */
+function adsenseRequestUrl(startTime, slotNumber, global, data, isAmp,
+                           canonicalUrl, correlator, intersectionRecord) {
+  // const lmt = global.document.lastModified;
+  const slot = intersectionRecord.boundingClientRect;
+  // For this intersectionRecord, root is the viewport.
+  const width = slot.width;
+  const height = slot.height;
+  const viewport = intersectionRecord.rootBounds;
+  const screen = global.screen;
+  // const fractionInViewport = intersectionRecord.intersectionRatio;
+  const visibilityState = documentStateFor(global).getVisibilityState();
+  const visNum = {
```

Comment:
I think this actually doesn't work. The information should be accessed through viewer service.

I'll file an issue for documenting this.